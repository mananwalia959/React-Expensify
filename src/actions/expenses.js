import uuid from 'uuid';


//ADD EXPENSE
export const addExpense = ({
    description='',
    note='',
    amount=0,
    createdAt=0,
    transaction="expense"
} ={}) => ({
    type:'AddExpense',
    expense:{
        id:uuid(),
        description,
        note,
        amount,
        createdAt,
        transaction
    }
});
//REMOVE EXPENSE
export const removeExpense=(( {id} )=> ({
    type:'remove',
    id
}))
//EDIT EXPENSE
export const editexpense=((id,updates)=>({
    type:'EditExpense',
    id,
    updates

}))