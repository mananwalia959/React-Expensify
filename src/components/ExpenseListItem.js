import React from 'react';
import {Link} from 'react-router-dom'



 const ExpenseListItem =({id,description,amount,createdAt,transaction}) => (
    <div>
    <Link to={`/edit/${id}`}>
    <h3>{transaction[0].toUpperCase()+ transaction.slice(1)}:{description}</h3>
    </Link>
    <p>{amount} - {createdAt}</p>
    </div>
)

export default ExpenseListItem;


//    <Link to={'/edit/'+id}>