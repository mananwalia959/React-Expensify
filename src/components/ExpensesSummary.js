import React from 'react';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom'
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expensestotal';

export const ExpensesSummary = ({expenseCount,transactiontotals}) => {
    const expenseWord = expenseCount === 1 ? "transaction":"transactions";
    const formattedExpenseTotal = (transactiontotals.total_expense/100)
    const formattedIncomeTotal   = (transactiontotals.total_income/100)


    return (
        <div className='page-header'>
         <div className='content-container'>
          <h1 className='page-header__title'> Viewing <span>{expenseCount}</span> {expenseWord} with Net Amount = <span>{formattedIncomeTotal-formattedExpenseTotal}</span> <br/>Expenses =<span>Rs.{formattedExpenseTotal}</span> and Incomes = <span>Rs.{formattedIncomeTotal}</span> </h1>
           <div>
            <Link className ='button' to='/create'>Add Transaction</Link>
           </div>
         </div>
        </div>
    )
}

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
  
    return {
      expenseCount: visibleExpenses.length,
      transactiontotals: selectExpensesTotal(visibleExpenses)
    };
  };
  
  export default connect(mapStateToProps)(ExpensesSummary);