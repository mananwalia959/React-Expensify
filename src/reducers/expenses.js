const expensesReducerDefaultState =[]

export default (state=expensesReducerDefaultState,action) => {
    switch(action.type){
        case 'AddExpense':
        return [
            ...state,
            action.expense
        ];
        case 'remove':
        return state.filter(({id})=> id !== action.id)

        case 'EditExpense':
        return state.map((expense)=>{
            if(expense.id===action.id){
                return({
                    ...expense,
                    ...action.updates})
            }
            else{
                return expense
            }
        })
        case 'SET_EXPENSES':
        return action.expenses;
        default:
        return state
        
    }
};