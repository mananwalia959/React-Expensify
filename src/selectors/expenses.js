import moment from 'moment';

export default (expenses,{text,sortby,startDate,endDate,transactiontype})=> {
    return expenses.filter((expense)=>{
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate?startDate.isSameOrBefore(createdAtMoment,'day'):true;
        const endDateMatch = endDate?endDate.isSameOrAfter(createdAtMoment,'day'):true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase()) ;
        const typeMatch= transactiontype==='both'?true:expense.transaction===transactiontype;

        return startDateMatch && endDateMatch && textMatch && typeMatch;
    }).sort((a,b)=> {
        if(sortby === 'date') {
            return a.createdAt < b.createdAt?1:-1;
        }
        if (sortby==='amount'){
            return a.amount<b.amount?1:-1
        }
    })
}