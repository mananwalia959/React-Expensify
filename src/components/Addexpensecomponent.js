import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {addExpense} from '../actions/expenses';
export class Addexpensecomponent extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/');

    }
    render() {
        return (
            <div>
            This is from my dashboard create component
            <ExpenseForm 
            onSubmit={this.onSubmit}
        />
    </div>

        )
    }
    
}

const mapDispatchToProps=(dispatch) => ({
    addExpense:(expense) => dispatch(addExpense(expense))
})


export default connect(undefined,mapDispatchToProps)(Addexpensecomponent);