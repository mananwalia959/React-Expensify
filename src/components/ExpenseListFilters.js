import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter, sortByAmount, sortByDate,setstartDate,setendDate,settransactiontype} from '../actions/filters'
import {DateRangePicker} from 'react-dates'

export class ExpenseListFilters extends React.Component{
    state={
        calendarFocused:null
    }
    onDatesChange=({startDate,endDate}) => {
        this.props.setstartDate(startDate);
        this.props.setendDate(endDate);
    };
    onFocusChange = (calendarFocused)=> {
        this.setState(()=>({calendarFocused}) )
    }
    onTextChange=(e) =>{
        this.props.setTextFilter(e.target.value)
    }
    onSortChange=(e) =>{
        if(e.target.value==='date'){
        this.props.sortByDate();
        }
        else if(e.target.value==='amount'){
        this.props.sortByAmount();
        }
    }
    onTransactionTypeChange=(e) =>{
        const option=e.target.value
        this.props.settransactiontype(option)
    }


    render(){
        return(
            <div>
        <input 
        type ='text' 
        value={this.props.filters.text} 
        onChange={this.onTextChange} 
        /> 

        <select 
        value={this.props.filters.sortby} 
        onChange={this.onSortChange}>
        <option value = 'date' > Date </option>
        <option value = 'amount'> Amount </option>
        </select>
        <select value={this.props.filters.transactiontype} 
        onChange={this.onTransactionTypeChange}>
        <option value = 'both' > Both </option>
        <option value = 'expense'> Expense </option>
        <option value = 'income' > Income </option>
        </select>
        <DateRangePicker
        startDate={this.props.filters.startDate}
        endDate={this.props.filters.endDate}
        onDatesChange={this.onDatesChange}
        focusedInput={this.state.calendarFocused}
        onFocusChange={this.onFocusChange}
        showClearDates={true}
        numberOfMonths={1}
        isOutsideRange={()=>false}
        />
        </div>
        )
        
    }
}
const MapStatetoProps=(state) => ({
    filters:state.filters
});

const MapDispatchToProps=(dispatch) => ({
    setTextFilter: (text)=> dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount:() => dispatch(sortByAmount()),
    setstartDate:(startDate)=>dispatch(setstartDate(startDate)),
    setendDate:(endDate)=> dispatch(setendDate(endDate)),
    settransactiontype:(tranaction)=> dispatch(settransactiontype(tranaction))
})


export default connect (MapStatetoProps,MapDispatchToProps)(ExpenseListFilters);
