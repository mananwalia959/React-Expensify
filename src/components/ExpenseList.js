import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import SelectExpenses from '../selectors/expenses';
export const ExpenseList = (props) => (
    <div className='content-container'>
      <div>
       <div className='list-header'>
        <div className='show-for-mobile'>Expenses</div>
        <div className='show-for-desktop'>Expense</div>
        <div className='show-for-desktop'>Amount</div>    
      </div>
    </div>


    <div className='list-body'>
    {
        props.expenses.length === 0?(
            <div className='list-item listitem--message'>
            <span>No Transactions</span>
            </div>
        ):(
            props.expenses.map((expense) => {
                return <ExpenseListItem key = {expense.id} {...expense}  />
            })
        )
    }
    </div> 
   
    </div>
)

const MapStateToProps = (state)=>{
    return {
        expenses:SelectExpenses(state.expenses,state.filters)
    }
}
export default connect(MapStateToProps)(ExpenseList);


