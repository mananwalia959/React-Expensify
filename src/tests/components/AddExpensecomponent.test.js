import React from 'react';
import {shallow} from 'enzyme';
import {Addexpensecomponent} from '../../components/Addexpensecomponent'
import expenses from '../fixtures/expenses';

let addExpense,history,wrapper ;

beforeEach(()=> {
    addExpense = jest.fn();
    history = {push: jest.fn()};
    wrapper = shallow(<Addexpensecomponent addExpense ={addExpense} history ={history} /> )
})


test('Should render AddExpensePage correctly',() => {
    expect(wrapper).toMatchSnapshot();
});

test('Should Handle onSubmit correctly',() => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(addExpense).toHaveBeenCalledWith(expenses[1]);

});