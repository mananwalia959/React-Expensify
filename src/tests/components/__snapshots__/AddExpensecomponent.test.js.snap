// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render AddExpensePage correctly 1`] = `
<div>
  This is from my dashboard create component
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
